### DISCUSSION TITLE
Tuple type with generic length creates errors with ts-node

### DISCUSSION THREAD

**Discussion by aurerua** [TypeStrong/ts-node#1350] [Q&A] [answered]
This following code:

```ts
type TupleOf<T, N extends number> = N extends N ? (number extends N ? T[] : _TupleOf<T, N, []>) : never;
type _TupleOf<T, N extends number, R extends unknown[]> = R["length"] extends N ? R : _TupleOf<T, N, [T, ...R]>;

type ThreeStrings = TupleOf<string,3>

const threeStrings = ['one', 'two', 'three']
```
works [in the TS Playground](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAKgVwgGygeQGYB44BoYByMUAHiFAHYAmAzjBQgLYBGUATgHwwC8hxZlWnwD8MABQMW7fuWp0iouAG0AujABcMAPqIU6bHkL5VHAJQb6UAG7sA3KEiwdSVJhz4ipWUMms2+ACUZQToECgBrCgB7AHcKEx4YAKUAIlQKAHMQAAsUtS8QkSSLZz03QwJjQwA6WoCVDnsHaHhstigoAGUQNgBLTLpeXVdsGh7+jNwAZg5gYABjKIoxmHG9RKUAciWoTfxNkBiovZgDto7NlSA) and shows no error in vscode but it creates errors when ts-node is run:

```
error TS2315: Type '_TupleOf' is not generic.

9 type TupleOf<T, N extends number> = N extends N ? (number extends N ? T[] : _TupleOf<T, N, []>) : never;
                                                                              ~~~~~~~~~~~~~~~~~~
10 type _TupleOf<T, N extends number, R extends unknown[]> = R["length"] extends N ? R : _TupleOf<T, N, [T, ...R]>;
        ~~~~~~~~
10 type _TupleOf<T, N extends number, R extends unknown[]> = R["length"] extends N ? R : _TupleOf<T, N, [T, ...R]>;
                                                                                         ~~~~~~~~~~~~~~~~~~~~~~~~~
``` 

Is there an easy explanation to that behavior?

**Comment by cspotcode** [comment, marked as answer]
Do the versions match?  `ts-node -vv`?

Does the behavior change with `ts-node --files` enabled?

**Reply by cspotcode** [reply]
Problem was using a different TS version on the playground than was installed locally.

**Comment by aurerua** [comment]
Thanks for the suggestions.

How are the versions supposed to match? Here is what I get:
```
$ ts-node -vv
ts-node v9.0.0
node v14.16.0
compiler v4.0.5
```
And `ts-node --files` does not help:
```ts
{
diagnostics: [
    {
      file: [SourceFileObject],
      start: 345,
      length: 18,
      messageText: "Type '_TupleOf' is not generic.",
      category: 1,
      code: 2315,
      reportsUnnecessary: undefined,
      reportsDeprecated: undefined
    },
    {
      file: [SourceFileObject],
      start: 379,
      length: 8,
      messageText: "Type alias '_TupleOf' circularly references itself.",
      category: 1,
      code: 2456,
      reportsUnnecessary: undefined,
      reportsDeprecated: undefined
    },
    {
      file: [SourceFileObject],
      start: 460,
      length: 25,
      messageText: "Type '_TupleOf' is not generic.",
      category: 1,
      code: 2315,
      reportsUnnecessary: undefined,
      reportsDeprecated: undefined
    }
  ]
}
```

**Reply by cspotcode** [reply]
What version are you using in the playground?

**Reply by cspotcode** [reply]
The playground shows the same thing for me: [Playground link](https://www.typescriptlang.org/play?ts=4.0.5#code/FAFwngDgpgBAKgVwgGygeQGYB44BoYByMUAHiFAHYAmAzjBQgLYBGUATgHwwC8hxZlWnwD8MABQMW7fuWp0iouAG0AujABcMAPqIU6bHkL5VHAJQb6UAG7sA3KEiwdSVJhz4ipWUMms2+ACUZQToECgBrCgB7AHcKEx4YAKUAIlQKAHMQAAsUtS8QkSSLZz03QwJjQwA6WoCVDnsHaHhstigoAGUQNgBLTLpeXVdsGh7+jNwAZg5gYABjKIoxmHG9RKUAciWoTfxNkBiovZgDto7NlSA)

**Reply by aurerua** [reply]
You are right. With version 4.0.5 in the playground this is the same issue. Thanks for you help.
